SymIntify first class dims

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101413
* #101409

To get pybind translation for SymInt, I just use the stock
pybind11 from PyTorch proper.  This is kind of annoying but
since first class dims are now in tree I don't think there is
too much benefit in forcing us to stay this way.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
SymIntify first class dims

To get pybind translation for SymInt, I just use the stock
pybind11 from PyTorch proper.  This is kind of annoying but
since first class dims are now in tree I don't think there is
too much benefit in forcing us to stay this way.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "SymIntify first class dims"


To get pybind translation for SymInt, I just use the stock
pybind11 from PyTorch proper.  This is kind of annoying but
since first class dims are now in tree I don't think there is
too much benefit in forcing us to stay this way.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

Skylion007 commented on functorch/csrc/dim/dim.cpp:
```diff
@@ -119,7 +121,7 @@ struct Dim : public mpy::base<Dim> {
-    void init(mpy::object name, int64_t s = -1) {
+    void init(mpy::object name, c10::SymInt s = -1) {
         name_ = std::move(name);
         size_ = s;
```
```suggestion
        size_ = std::move(s);
```

Skylion007 commented on functorch/csrc/dim/dim.cpp:
```diff
@@ -128,17 +130,17 @@ struct Dim : public mpy::base<Dim> {
-    void set_size(int64_t v) {
+    void set_size(c10::SymInt v) {
         if (size_ == -1) {
             size_ = v;
```
```suggestion
            size_ = std::move(v);
```

ezyang commented:
There is something fundamentally incomplete about this PR, which is the interaction of arena with SymInt alllocation. IIUC, the arena is for fast allocation of non-owning chunks of data, which used to be OK with int, but is no longer OK with SymInt. I think probably what we want to do is have the arena hold onto the SymInts that are "in scope" (similar to its autorelease on "Tensor") but I'm not entirely clear how to do this, or if it is right. Need some input from @zdevito

ezyang committed:
Update on "SymIntify first class dims"


To get pybind translation for SymInt, I just use the stock
pybind11 from PyTorch proper.  This is kind of annoying but
since first class dims are now in tree I don't think there is
too much benefit in forcing us to stay this way.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
